src/main/java/org/zalando/nakadi/service/PaginationService.java

```diff
@@ -0,0 +1,68 @@
+package org.zalando.nakadi.service;
+
+import org.springframework.stereotype.Component;
+import org.springframework.web.util.UriComponentsBuilder;
+import org.zalando.nakadi.domain.PaginationLinks;
+import org.zalando.nakadi.domain.PaginationWrapper;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+@Component
+public class PaginationService {
+
+    public PaginationWrapper paginate(final int offset,
+                                      final int limit,
+                                      final String path,
+                                      final ItemsSupplier itemsSupplier,
+                                      final Supplier<Integer> countSupplier) {
+        final List items = itemsSupplier.queryOneMore(offset, limit);
+        final PaginationLinks paginationLinks;
+        if (items.isEmpty() && offset != 0 && limit != 0) {
+            final int count = countSupplier.get();
+            int latestOffset = count / limit;
+            if (offset >= latestOffset)
+                latestOffset = 0;
+            paginationLinks = createLinks(path, latestOffset, limit, items.size());
+        } else {
+            paginationLinks = createLinks(path, offset, limit, items.size());
+            if (items.size() > limit)
```

Comment:
For example you have 10 elements, offset=0, limit=10. How can you identify that you should or should not show to show the link to the next page?
My approach is to request one more element from db to understand if we have more elements to show on the next page.